services/newsfeed/tests/unit/shared/twitter.test.js

```diff
@@ -0,0 +1,205 @@
+		const url = getImageUrl(tweetObject);
+		expect(url).toBe(undefined);
+	});
+
+	it('should return undefined when called with retweet', async () => {
+		const url = getImageUrl(retweetObject);
+		expect(url).toBe(undefined);
+	});
+
+	it('should return image url when called with mediaTweet', async () => {
+		const url = getImageUrl(mediaTweetObject);
+		expect(url).toBe(mediaTweetObject.entities.media[0].media_url_https);
+	});
+
+	it('should return undefined when called with otherTweet', async () => {
+		const url = getImageUrl(otherTweetObject);
+		expect(url).toBe(undefined);
+	});
+
+	it('should return correct url when called with an object having entities.media[0].media_url_https', async () => {
+		const obj = {
+			entities: {
+				media: [{
+					media_url_https: 'expected_url',
+				}],
+			},
+		};
+		const url = getImageUrl(obj);
+		expect(url).toBe(obj.entities.media[0].media_url_https);
+	});
+
+	it('should throw error when called with a null object', async () => {
+		expect(() => getImageUrl(null)).toThrow();
+	});
+
+	it('should throw error when called with a undefined object', async () => {
+		expect(() => getImageUrl(undefined)).toThrow();
+	});
+});
+
+describe('Test getTweetText method', () => {
+	it('should return tweet.text when called with tweet', async () => {
+		const url = getTweetText(tweetObject);
+		expect(url).toBe(tweetObject.text);
+	});
+
+	it('should return retweet.text when called with retweet', async () => {
+		const url = getTweetText(retweetObject);
+		expect(url).toBe(retweetObject.text);
+	});
+
+	it('should return mediaTweet.text when called with mediaTweet', async () => {
+		const url = getTweetText(mediaTweetObject);
+		expect(url).toBe(mediaTweetObject.text);
+	});
+
+	it('should return otherTweet.text when called with otherTweet', async () => {
+		const url = getTweetText(otherTweetObject);
+		expect(url).toBe(otherTweetObject.text);
+	});
+
+	it('should return text with quote information when called with an object having is_quote_status:true and quoted_status', async () => {
+		const obj = {
+			text: 'custom tweet text',
+			is_quote_status: true,
+			quoted_status: {
+				text: 'custom_quoted_text',
+				user: {
+					name: 'user_name',
+					screen_name: 'user_screen_name',
+				},
+			},
+		};
+		const expectedText = `${obj.text}\n\nQuoted status by ${obj.quoted_status.user.name} (@${obj.quoted_status.user.screen_name}): ${obj.quoted_status.text}`;
+		const url = getTweetText(obj);
+		expect(url).toBe(expectedText);
+	});
+
+	it('should throw error when called with null object', async () => {
+		expect(() => getTweetText(null)).toThrow();
+	});
+
+	it('should throw error when called with undefined object', async () => {
+		expect(() => getTweetText(undefined)).toThrow();
+	});
+});
+
+describe('Test tweetUrl method', () => {
+	it('should return correct url when called with tweet', async () => {
+		jest.spyOn(twitterMethods, 'safeRef').mockReturnValueOnce(tweetObject.entities.urls[0].url);
```

Comment:
I think the spy used here is not really mocking the response. Can you re-verify?

Reply:
Seems to be working on my system.

Reply:
Yes, the tests are passing but the concern here is jest.spyOn() is not really mocking the function instead real implemented is being executed. Tests will pass even with removing L171. So, we can remove all mocks here.

Reply:
Yes, I retested it and seems like the spy isn't working. @priojeetpriyom please remove the spy in all the testcases.

Reply:
Moving the test under functional. Thanks for noticing guys 🙏

Reply:
Yeah, it is dependant on `safeRef` which we were trying to mock.
https://github.com/LiskHQ/lisk-service/blob/1471-add-unit-tests-for-newsfeed-microservice/services/newsfeed/shared/twitter.js#L48-L52

Reply:
`safeRef` isn't dependant on any external service. Please move the tests back to unit. It should be fine.

Reply:
Reverted.